### PR TITLE
Add `assertDoesNotAddMessages` to `CheckerTestCase`

### LINKED ISSUE #9598: Allow CheckerTestCase to assertDoesNotAddMessages() to check that a specific message has not been added
### Current problem

I use `CheckerTestCase` in tests for linting scripts in a project. There are some tests where I want to assert that a specific message has not been added given a specific code snippet. While `assertNoMessages()` works for now, it is possible in the future that I want to assert that a different message occurs for the same code snippet. `assertNoMessages()` wouldn't work because we do want to check that the different message has been added, but replacing `assertNoMessages()` with `assertAddsMessages(different_message, ...)` would mean that we can no longer check that the original message is not being added.  

### Desired solution

If I could directly assert that a specific message has not been added, then I could have more control over which specific messages I'd want or would not want under various scenarios. For my given example, something like this:
```python
assertDoesNotAddMessages(original_message, ...)
assertAddsMessages(different_message, ...)
```
Better represents which messages I want to occur and which messages I want to not occur.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ShehabSherif0**
## Summary

Adds `assertDoesNotAddMessages(*message_ids: str)` to `CheckerTestCase` as a context manager that asserts specific message IDs are not emitted during a checker run, while allowing other messages to be present.

This fills the gap between `assertNoMessages()` (no messages at all) and `assertAddsMessages()` (exact set of messages). Now you can write tests like:

```python
with self.assertDoesNotAddMessages("missing-module-docstring"):
    self.checker.visit_module(node)
```

## Motivation

Closes #9598. When testing checkers, it is useful to verify a specific message is not raised for a given input without requiring that zero messages are raised. For example, confirming a checker no longer emits a false positive for a pattern, even if it still correctly emits other warnings.

## Changes

- `pylint/testutils/checker_test_case.py`: Added `assertDoesNotAddMessages(*message_ids: str)` context manager.
- `tests/testutils/test_checker_test_case.py`: New test file covering all six scenarios from the issue plus edge cases.
- `doc/whatsnew/fragments/9598.internal`: Towncrier news fragment.

## Test Scenarios

Following the scenarios from @Pierre-Sassoulas in the issue:

| # | Scenario | Method | Expected |
|---|----------|--------|----------|
| 1 | expected raised / actual raised | `assertAddsMessages` | pass |
| 2 | expected raised / actual not raised | `assertAddsMessages` | fail |
| 3 | expected raised / actual not raised but another raised | `assertAddsMessages` | fail |
| 4 | expected not raised / actual raised | `assertDoesNotAddMessages` | fail |
| 5 | expected not raised / actual not raised | `assertDoesNotAddMessages` | pass |
| 6 | expected not raised / actual not raised but another raised | `assertDoesNotAddMessages` | pass |

Plus edge cases: no arguments raises `TypeError`, multiple unwanted IDs fail if any match, and exceptions inside the body drain the message buffer cleanly.

### CODE REVIEW

File: doc/whatsnew/fragments/9598.feature [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Add ``assertDoesNotAddMessages`` to ``CheckerTestCase`` to assert that specific messages are not emitted, while allowing other messages to be present. This complements ``assertNoMessages`` which asserts that no messages at all are emitted.
+
+Closes #9598
```
[the remark this answers is not in the document]
Comment (ShehabSherif0) [reply doc/whatsnew/fragments/9598.feature]: Valid, fixed both. Switched to `Refs #9598` and wrapped the long first line to match the format used in other fragments.

File: pylint/testutils/checker_test_case.py [hunk 2]
```diff
@@ -37,6 +37,54 @@ def assertNoMessages(self) -> Iterator[None]:
...
             yield
 
+    @contextlib.contextmanager
+    def assertDoesNotAddMessages(
+        self, *messages: MessageTest, ignore_position: bool = False
+    ) -> Generator[None]:
+        """Assert that the given messages are not added by the given method.
+
+        This is different from ``assertNoMessages`` which asserts that no
+        messages at all are added.  ``assertDoesNotAddMessages`` checks that
```
[the remark this answers is not in the document]
Comment (ShehabSherif0) [reply pylint/testutils/checker_test_case.py]: Valid, fixed.

File: pylint/testutils/checker_test_case.py [hunk 3]
```diff
@@ -37,6 +37,54 @@ def assertNoMessages(self) -> Iterator[None]:
         with self.assertAddsMessages():
             yield
 
+    @contextlib.contextmanager
+    def assertDoesNotAddMessages(
+        self, *messages: MessageTest, ignore_position: bool = False
+    ) -> Generator[None]:
```
[the remark this answers is not in the document]
Comment (ShehabSherif0) [reply pylint/testutils/checker_test_case.py]: Valid. Fixed by raising `TypeError("assertDoesNotAddMessages requires at least one MessageTest argument")` at the top of the method. Added a test to cover this case as well.

File: pylint/testutils/checker_test_case.py [hunk 4]
```diff
@@ -37,6 +37,54 @@ def assertNoMessages(self) -> Iterator[None]:
...
+            for gotten_msg in got:
+                if not self._messages_match(unwanted, gotten_msg, ignore_position):
+                    continue
+                got_str = "\n".join(repr(m) for m in got)
+                msg = (
+                    "Expected the following message to not be raised:\n"
+                    f"\n  {unwanted!r}\n\n"
+                    f"but it was found among the actual messages:\n\n{got_str}\n"
+                )
+                raise AssertionError(msg)
```
[the remark this answers is not in the document]
Comment (ShehabSherif0) [reply pylint/testutils/checker_test_case.py]: Valid concern, fixed. I went with `try/except/else` rather than the suggested `try/finally` because the suggestion has a subtle bug: calling `return` inside a `finally` block suppresses the in-flight exception in Python. That means if the test body raises (e.g. an `AttributeError`), the original error would be silently swallowed. With `try/except/else`, the `except` branch drains messages and re-raises cleanly, and the `else` branch only runs when no exception occurred.

File: doc/whatsnew/fragments/9598.internal [hunk 5]
```diff
@@ -0,0 +1,6 @@
+Add ``assertDoesNotAddMessages`` to ``CheckerTestCase`` to assert that
```
Comment (Pierre-Sassoulas) [inline doc/whatsnew/fragments/9598.internal:1]: I think it should be "internal" as it touches the API for pylint plugin developpers

File: pylint/testutils/checker_test_case.py [hunk 6]
```diff
@@ -37,6 +37,63 @@ def assertNoMessages(self) -> Iterator[None]:
...
+        if not ignore_position:
+            if expected.line != actual.line:
+                return False
+            if expected.col_offset != actual.col_offset:
+                return False
+            if expected.end_line != actual.end_line:
+                return False
+            if expected.end_col_offset != actual.end_col_offset:
+                return False
+        return True
```
Comment (Pierre-Sassoulas) [inline pylint/testutils/checker_test_case.py]: Don't we only care about the msgid ?

**Review by Pierre-Sassoulas — COMMENTED** [review]
Thank you for contributing to pylint. Sorry fr the delay I didn't send the review I thought I sent.

File: doc/whatsnew/fragments/9598.internal (hunk 5 above)
Comment (ShehabSherif0) [reply doc/whatsnew/fragments/9598.internal:1]: Good catch, renamed to `9598.internal`.

File: pylint/testutils/checker_test_case.py (hunk 6 above)
Comment (ShehabSherif0) [reply pylint/testutils/checker_test_case.py]: You are right. The method now accepts plain message ID strings instead of full `MessageTest` objects, matching the intended use pattern from the issue. The check is just: if that message ID appears anywhere in the emitted messages, fail. Position, args, node and confidence are all irrelevant to the "did this message fire at all" question. The `_messages_match` helper and `ignore_position` parameter have been removed.

File: tests/testutils/test_checker_test_case.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+
+_MSG_A = MessageTest("W9901", line=1, node=None, args=None, confidence=UNDEFINED)
+
+
+class TestCheckerTestCase(CheckerTestCase):
+    CHECKER_CLASS = _DummyChecker
+
+    # -- assertAddsMessages scenarios (1-3) -----------------------------------
```
Comment (Pierre-Sassoulas) [inline tests/testutils/test_checker_test_case.py]:
```suggestion
```

File: tests/testutils/test_checker_test_case.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+            with self.assertAddsMessages(_MSG_A):
+                pass  # nothing emitted
+
+    def test_assert_adds_messages_failure_wrong_message(self) -> None:
+        """Scenario 3: expected raised / actual not raised but another one raised."""
+        with pytest.raises(AssertionError):
+            with self.assertAddsMessages(_MSG_A):
+                self.linter.add_message("W9902", line=2)
+
+    # -- assertDoesNotAddMessages scenarios (4-6) -----------------------------
```
Comment (Pierre-Sassoulas) [inline tests/testutils/test_checker_test_case.py]:
```suggestion
```

File: tests/testutils/test_checker_test_case.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        """Scenario 5: expected not raised / actual not raised."""
+        with self.assertDoesNotAddMessages("W9901"):
+            pass  # nothing emitted
+
+    def test_assert_does_not_add_messages_success_other_raised(self) -> None:
+        """Scenario 6: expected not raised / actual not raised but another one raised."""
+        with self.assertDoesNotAddMessages("W9901"):
+            self.linter.add_message("W9902", line=2)
+
+    # -- additional edge cases ------------------------------------------------
```
Comment (Pierre-Sassoulas) [inline tests/testutils/test_checker_test_case.py]:
```suggestion
```

File: tests/testutils/test_checker_test_case.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+    def test_assert_does_not_add_messages_no_args_raises(self) -> None:
+        """Calling with no arguments must raise TypeError."""
+        with pytest.raises(TypeError, match="requires at least one"):
+            with self.assertDoesNotAddMessages():
+                pass
+
+    def test_assert_does_not_add_messages_multiple_unwanted(self) -> None:
+        """Fails when any of the several unwanted message IDs is found."""
+        with pytest.raises(AssertionError):
```
Comment (Pierre-Sassoulas) [inline tests/testutils/test_checker_test_case.py]:
```suggestion
        with pytest.raises(AssertionError, match="..."):
```

File: tests/testutils/test_checker_test_case.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        """Scenario 3: expected raised / actual not raised but another one raised."""
+        with pytest.raises(AssertionError):
+            with self.assertAddsMessages(_MSG_A):
+                self.linter.add_message("W9902", line=2)
+
+    # -- assertDoesNotAddMessages scenarios (4-6) -----------------------------
+
+    def test_assert_does_not_add_messages_failure(self) -> None:
+        """Scenario 4: expected not raised / actual raised."""
+        with pytest.raises(AssertionError):
```
Comment (Pierre-Sassoulas) [inline tests/testutils/test_checker_test_case.py]:
```suggestion
        with pytest.raises(AssertionError, match="..."):
```

File: tests/testutils/test_checker_test_case.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    # -- assertAddsMessages scenarios (1-3) -----------------------------------
+
+    def test_assert_adds_messages_success(self) -> None:
+        """Scenario 1: expected raised / actual raised."""
+        with self.assertAddsMessages(_MSG_A):
+            self.linter.add_message("W9901", line=1)
+
+    def test_assert_adds_messages_failure_not_raised(self) -> None:
+        """Scenario 2: expected raised / actual not raised."""
+        with pytest.raises(AssertionError):
```
Comment (Pierre-Sassoulas) [inline tests/testutils/test_checker_test_case.py]:
```suggestion
        with pytest.raises(AssertionError, match="..."):
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Thank you, let's add some assertion about the message raised and let's merge.

**Review by Pierre-Sassoulas — APPROVED** [review]
Thank you for your work on pylint :) This is going in 4.1.0 !
